### DISCUSSION TITLE
Step not doable on my machine, how to work around it?

### DISCUSSION THREAD

**Discussion by llrs** [ropensci/targets#649] [Help]
I have a step of an analysis that I cannot calculate on my machine. I have access to a server where I can do that calculation.
I have done that step on the server but I want to continue using the results of it on my local machine.  

I would leave a comment but it won't show up as needing to update on `tar_visnetwork()`  nor would `tar_make()` be able to compute that step if I include it on the workflow directly. 

What is the best way to flag that on targets? If I include the data on a file (the end result is not big, it is just 3 numbers) it won't be updated, or can i just copy the accompanying _targets/meta/*?

In pseudo-code:
```r
list(
    tar_target(data, functions())
    tar_target(hard_step, calculate_numbers(data)) # Doesn't work on my local machine but does on the server
    tar_target(keep_workflow, analysis_part2(hard_step))
)
```

**Comment by wlandau-lilly** [comment]
Hi Lluís!

You could copy all of `_targets/` to your local machine and then force that target to never run using a cue:

```r
# _targets.R
library(targets)
on_server <- function() {
  # check if on server
}
list(
  tar_target(data, functions()),
  tar_target(hard_step, calculate_numbers(data), cue = tar_cue(mode = ifelse(on_server(), "thorough", "never"))),
  tar_target(keep_workflow, analysis_part2(hard_step))
)
```

**Reply by llrs** [reply]
Hi Will! 

Sorry, first time using targets on a project and just recently read about the cues.
If cue is set to "never" (on my local machine) and I change/update upstream targets would it show as not updated on my local machine via `tar_visnetwork()` or `tar_outdated`? 

----

The server doesn't have any browser so when I tried to use `tar_visnetwork` I received :

```
> tar_visnetwork()
Couldn't find a suitable web browser!
Set the BROWSER environment variable to your desired browser.
Warning: program returned non-zero exit code #256
Opening "/tmp/RtmppTQv4Q/viewhtml398815fd2a01f/index.html" with Vim  (text/html)
Can't exec "xterm": No such file or directory at /usr/share/perl5/File/DesktopEntry.pm line 247.
/usr/bin/xdg-open: 869: x-www-browser: not found
/usr/bin/xdg-open: 869: firefox: not found
/usr/bin/xdg-open: 869: iceweasel: not found
/usr/bin/xdg-open: 869: seamonkey: not found
/usr/bin/xdg-open: 869: mozilla: not found
/usr/bin/xdg-open: 869: epiphany: not found
/usr/bin/xdg-open: 869: konqueror: not found
/usr/bin/xdg-open: 869: chromium: not found
/usr/bin/xdg-open: 869: chromium-browser: not found
/usr/bin/xdg-open: 869: google-chrome: not found
/usr/bin/xdg-open: 869: www-browser: not found
/usr/bin/xdg-open: 869: links2: not found
/usr/bin/xdg-open: 869: elinks: not found
/usr/bin/xdg-open: 869: links: not found
/usr/bin/xdg-open: 869: lynx: not found
/usr/bin/xdg-open: 869: w3m: not found
xdg-open: no method available for opening '/tmp/RtmppTQv4Q/viewhtml398815fd2a01f/index.html'
```

It would be nice if this were handled more graciously or providing a static image if no browser is found.

**Comment by wlandau** [comment]
> If cue is set to "never" (on my local machine) and I change/update upstream targets would it show as not updated on my local machine via tar_visnetwork() or tar_outdated?

`tar_cue(mode = "never")` makes sure the target only runs if a return value does not exist in storage. The build behavior will not respond to changes in upstream targets. Downstream targets, however, will respond.

> It would be nice if this were handled more graciously or providing a static image if no browser is found.

`targets` uses the `visNetwork` package for the graphing, so that is outside my control, unfortunately.

**Comment by llrs** [comment]
If I understood you correctly that would make moot the target (and the whole usage of targets) as I would need to remember if I changed the upstream or not on my local machine. I could work on the server but the Rstudio Server/Workbench is using a different R version than my local machine (there are other projects). 

It seems that any target that depends on a target with mode never is skipped. Isn't there an intermediate mode? Something like: Don't run but report that it is not updated (and the same with downstream targets)?

Thanks for the quick replies!

**Reply by wlandau** [reply]
> Something like: Don't run but report that it is not updated (and the same with downstream targets)?

Not sure I follow, but there is no way to programmatically propagate cues just to downstream targets specifically. Maybe just work on the server and use your local machine only for visualization/EDA?

**Reply by llrs** [reply]
I'm suggesting to have:

```r
# _targets.R
library(targets)
on_server <- function() {
  # check if on server
}
list(
  tar_target(data, functions()),
  tar_target(hard_step, calculate_numbers(data), cue = tar_cue(mode = "partial"))),
  tar_target(keep_workflow, analysis_part2(hard_step))
)
```

Where partial mode would mark the target as outdated if any upstream target changed but wouldn't run the target. 
Now that I think about it I could move to the server without much trouble, but I would prefer to work on the local machine.

**Reply by wlandau** [reply]
Sorry, partial mode does not seem like a good fit for the intent of the tool, and I do not think it would be used very often.

**Reply by llrs** [reply]
Ok. Many thanks for all the useful feedback and developing and maintaining the package!
